src/security/token/TKTokenCryptoEngine.m

```diff
@@ -48,15 +53,74 @@ - (TKSignature *)signData:(NSData *)data
             usingKeyLevel:(Key_Level)keyLevel
                    reason:(NSString *)reason
                   onError:(OnError)onError {
+    if (!useLocalAuthentication) {
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    if (keyLevel >= Key_Level_Low) {
+        // We don't check DeviceOwnerAuthentication for Key_Level_Low
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    LAContext *context = [[LAContext alloc] init];
+    
+    if (![context canEvaluatePolicy:LAPolicyDeviceOwnerAuthentication error:nil]) {
+        // If device don't support Device Owner Authentication, skip as success
+        return [self createSignature:data usingKeyLevel:keyLevel];
+    }
+    
+    dispatch_semaphore_t semaphore = dispatch_semaphore_create(0);
+    __block TKSignature *signature = nil;
+    __block NSError *laError = nil;
+    
+    [context evaluatePolicy:LAPolicyDeviceOwnerAuthentication
+            localizedReason:reason
+                      reply:^(BOOL success, NSError *error) {
+                          // This is a backend thread
```

Comment:
I don't understand this comment.

Reply:
I will change this to "This is not a main thread".
This is important, so that we can use semaphore without manually run the runloop.